### PR TITLE
feat(eks): Allow helm pull from OCI repositories

### PR DESCRIPTION
**Pull request by harshadbhatia**
The feature allows lambda to install charts from OCI repositories. This also adds login capabilities when the AWS registry is used.

Fixes - https://github.com/aws/aws-cdk/issues/18001

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-eks/lib/kubectl-handler/helm/__init__.py [hunk 1]
```diff
@@ -78,13 +80,72 @@ def helm_handler(event, context):
...
+
+    cmnd = []
+    pattern = '\d+.dkr.ecr.[a-z]+-[a-z]+-\d.amazonaws.com'
+
+    registry = repository.rsplit('/', 1)[0].replace('oci://', '')
+
+    if re.fullmatch(pattern, registry) is not None:
+        region = registry.replace('.amazonaws.com', '').split('.')[-1]
+        cmnd = [
+            "HELM_EXPERIMENTAL_OCI=1;" \
```
[the remark this answers is not in the document]
Comment (harshadbhatia) [reply packages/@aws-cdk/aws-eks/lib/kubectl-handler/helm/__init__.py:107]: @otaviomacedo , I think that could be removed. Let me test and push the change.

[the remark this answers is not in the document]
Comment (harshadbhatia) [reply packages/@aws-cdk/aws-eks/lib/kubectl-handler/helm/__init__.py]: @otaviomacedo , added README and removed from above.

**Comment by harshadbhatia** [comment]
@otaviomacedo , Any updates on this ?

File: packages/@aws-cdk/aws-eks/README.md [hunk 2]
```diff
@@ -1144,6 +1144,24 @@ cluster.addHelmChart('test-chart', {
 });
 ```
 
+### OCI Charts
+
+OCI charts are also supported. Before executing ensure the handler lambda has the required ECR IAM permissions.
```
Comment (otaviomacedo) [inline packages/@aws-cdk/aws-eks/README.md:1149]: What are the required permissions? Can we automate that for the user?

**Comment by harshadbhatia** [comment]
They would be ecr repository permissions to be able to log in. Ideally, it
should be present on the lambda.

Not sure where lambda IAM Is present to verify the same.

File: packages/@aws-cdk/aws-eks/README.md (hunk 2 above)
Comment (harshadbhatia) [reply packages/@aws-cdk/aws-eks/README.md:1149]: They would be ecr repository permissions to be able to log in. Ideally, it
should be present on the lambda.

Comment (otaviomacedo) [reply packages/@aws-cdk/aws-eks/README.md]: The handler role is this one:

https://github.com/aws/aws-cdk/blob/19a287f02bc427644837956466213ee65457a857/packages/@aws-cdk/aws-eks/lib/kubectl-provider.ts#L164

You will need to add the `ecr:GetAuthorizationToken` permission to it

Comment (harshadbhatia) [reply packages/@aws-cdk/aws-eks/README.md]: @otaviomacedo , Done. please review

**Review by otaviomacedo — DISMISSED** [review]

File: packages/@aws-cdk/aws-eks/README.md (hunk 2 above)
Comment (otaviomacedo) [reply packages/@aws-cdk/aws-eks/README.md]: Looks good. But the tests have to be updated now.

Comment (harshadbhatia) [reply packages/@aws-cdk/aws-eks/README.md]: @otaviomacedo , I tried to look for a test under cdk-eks. Cant see anything. I assumed there are no tests for the same. Neither the documentation tells how all of this is actually tested.

If you can point me to the location. I am happy to write tests.

Comment (harshadbhatia) [reply packages/@aws-cdk/aws-eks/README.md]: @otaviomacedo , When I am trying to run tests from aws-eks directory. It appears to be not picking up tests at all or giving other errors. Could you advise me on the correct steps to execute tests?

Comment (otaviomacedo) [reply packages/@aws-cdk/aws-eks/README.md]: `yarn build && yarn test` should be the only thing you need. Another thing you can try is to build the whole transitive closure before running the test:

```sh
 aws-eks $ ../../../scripts/buildup
```

Comment (harshadbhatia) [reply packages/@aws-cdk/aws-eks/README.md]: @otaviomacedo , the tests are now complete. Could we please merge this ?

Comment (hbhatia-demyst) [reply packages/@aws-cdk/aws-eks/README.md]: @, Could you please review. Thanks

**Review by otaviomacedo — APPROVED** [review]

**Review by otaviomacedo — APPROVED** [review]

**Comment by harshadbhatia** [comment]
@otaviomacedo , thanks a lot for your help !

**Comment by tsahiduek** [comment]
Hi, Is this available at CDK V2 2.19.0?
How should I use it?
I've tried the following but it fails:
```ts
    cluster.addHelmChart(
      'AckS3', {
      chart: 'ack-s3',
      repository: 'oci://public.ecr.aws/aws-controllers-k8s/s3-chart',
      version: 'v0.0.19',
      namespace: 'ack-system',
      createNamespace: true,
    });
```

TIA

**Comment by tsahiduek** [comment]
Did someone tested this feature? 
I've copied the python code locally to understand why it's not working, and it seems that the [`helm pull`](https://github.com/harshadbhatia/aws-cdk/blob/master/packages/@aws-cdk/aws-eks/lib/kubectl-handler/helm/__init__.py#L112) command is executed inside the subprocess, but it doesn't really working. When tested locally the output directory is empty (I've modified the `tempfile.TemporaryDirectory()` creation to `tempfile.mkdtemp()` in order to persist the directory after the program finished to run

Any help here will be much appreciated

**Comment by tsahiduek** [comment]
I think I've understood why:
The [`check_output`](https://stackoverflow.com/questions/3172470/actual-meaning-of-shell-true-in-subprocess) uses `shell=True`. That means that all arguments of the commands being passed to the `check_output` are basically been passed  to the shell and not to the `helm pull` command. Using shell is also discouraged from [security perspective](https://docs.python.org/3/library/subprocess.html#security-considerations)
References:
https://docs.python.org/3/library/subprocess.html
https://stackoverflow.com/questions/3172470/actual-meaning-of-shell-true-in-subprocess

I'll open a PR to fix that (tested locally)

**Pull request merged**